File: OrcJava/src/orc/Orc.java
//
// Orc.java -- Java class Orc
// Project OrcJava
//
// $Id$
//
// Copyright (c) 2010 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc;

import java.io.IOException;

import orc.ast.oil.expression.Expression;
import orc.error.compiletime.CompileMessageRecorder.Severity;
import orc.runtime.OrcEngine;

/**
 * Main class for Orc. Parses Orc file and executes it.
 * <p>
 * Run with the argument "-help" to get a list of command-line options.
 * 
 * @author wcook, dkitchin
 */
public class Orc {

	/**
	 * Orc toplevel main function. Command line arguments are forwarded to Config for parsing.
	 * 
	 * @param args command-line arguments
	 */
	public static void main(final String[] args) {

		// Read configuration options from the environment and the command line
		final Config cfg = new Config();
		cfg.processArgs(args);

		final OrcCompiler compiler = new OrcCompiler(cfg);
		Expression ex;
		try {
			ex = compiler.call();
		} catch (final IOException e) {
			cfg.getMessageRecorder().recordMessage(Severity.FATAL, 0, e.getLocalizedMessage(), null, null, e);
			return;
		}

		if (ex == null) {
			return;
		}

		
		if (cfg.noExecute()) {
			System.out.println("Execution suppressed by -noexecute switch.");
		}
		else {
			// Configure the runtime engine
			final OrcEngine engine = new OrcEngine(cfg);

			// Run the Orc program
			engine.run(ex);
		}
	}

	public static Expression compile(final Config cfg) throws IOException {
		final OrcCompiler compiler = new OrcCompiler(cfg);
		return compiler.call();
	}
}

File: branches/PetersOrcMachine/OrcScala/src/orc/run/Orc.scala
//
// Orc.scala -- Scala class Orc
// Project OrcScala
//
// $Id$
//
// Created by dkitchin on May 10, 2010.
//
// Copyright (c) 2012 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc.run

import orc.ast.oil.nameless.Expression
import orc.{ OrcRuntime, OrcExecutionOptions, OrcEvent }
import orc.run.core.{ Token, Execution }
import scala.collection.mutable.SynchronizedSet
import scala.collection.mutable.HashSet
import scala.ref.WeakReference

abstract class Orc(val engineInstanceName: String) extends OrcRuntime {

  var roots: HashSet[WeakReference[Execution]] = new HashSet[WeakReference[Execution]] with SynchronizedSet[WeakReference[Execution]]

  def run(node: Expression, k: OrcEvent => Unit, options: OrcExecutionOptions) {
    startScheduler(options: OrcExecutionOptions)

    val root = new Execution(node, options, k, this)
    installHandlers(root)

    roots += new WeakReference(root)

    val t = new Token(node, root)
    schedule(t)
  }

  def stop() = {
    stopScheduler()
  }

  /** Add all needed event handlers to an execution.
    * Traits which add support for more events will override this
    * method and introduce more handlers.
    */
  def installHandlers(host: Execution) {}

}